content/macro-annotations.md

```diff
@@ -0,0 +1,349 @@
+  cache(n)
+```
+
+### Example 3: Main annotation
+
+In this example, we define an `@httpMain` that will generate a `main` method in a similar way to the `@main` method in Scala 3.
+
+
+```scala
+@httpMain def httpGet(url: URL) = println(http.get(url))
+```
+
+The `@httpMain` will generate a new class named `httpMain` with a `main` method that will call the annotated definition with the parsed arguments. Note that the syntax below is not valid Scala, but a representation of the TASTy code that would be generated. In particular, the top-level definition is moved into its package object `Example3$package`, and the
+`main` method is generated in a class as a static method rather than an object.
+
+```scala
+object Example3$package {
+  def httpGet(url: URL) = println(http.get(url))
+}
+
+class httpGet {
+  <static> def main(args: Array[String]): Unit =
+    args match
+      case Array(urlString) => Example3$package.httpGet(URL.parse(url))
+      case _ => println("Usage: httpGet <url>")
+}
+```
+
+## Design constraints
+
+Macro annotations must integrate with existing language features.
+
+* Integrate with existing macro system and safety guarantees
+* Separate/incremental compilation
+* IDE support
+
+### Integrate with existing macro system and safety guarantees
+We want seamless integration with the existing macro system. This means that macro annotations should be able to use the reflection API and the multi-stage programming abstractions (quoted expressions  `'{...}`). This also implies that we do not need to invent yet another new metaprogramming API for macro annotations.
+
+### Separate/incremental compilation
+Supporting incremental compilation implies that macro annotations must not affect the typing process.
+Otherwise, the order in which files are compiled/typed would introduce non-deterministic typing outcomes.
+
+### IDE/tools support
+For this, we rely on the representation of the program in TASTy. The IDEs or other tools should not get confused by changes in definitions caused by macro annotations. In other words, the TASTy representation of the program must contain all the information of the original definitions.
+
+
+## Proposed solution
+
+The library will define a macro annotation trait that will define a tree transformation operation using the `scala.quotes.Quotes` API.
+A macro implementation is a class that extends this trait and implements the `transform` method.
+Then a definition annotated with the concrete macro annotation will be transformed by the macro implementation at compile time.
+
+### High-level overview
+
+The `MacroAnnotation` annotation trait would be defined as follows:
+
+```scala
+package scala.annotation
+import scala.quoted.*
+
+/** Annotation classes that extend this trait will be able to transform the annotated definition. */
+trait MacroAnnotation extends StaticAnnotation:
+  /** Transforms a tree `definition` and optionally adds new definitions.
```

Comment:
Cask and MainArgs are perhaps one of the more common libraries using "fake" macro annotations today, and I suspect they wouldn't be implementable using this current proposal as is

That because they require that multiple `@annotated` methods be considered to generate a single combined routing object. In MainArgs you can have multiple `@main` methods, and in Cask you can have multiple `@cask.get`/`@cask.post`/etc. routes. 

In both cases, we currently have a expression-macro that looks at the class body, looks for all the `@annotated def`s within it, and generates a single expression that references all of them.

Is there some way we can support that sort of use case?

Reply:
maybe an aggregate version that only returns new trees to go at the end of the surrounding scope?

Reply:
Maybe we could have a macro annotation `@flask`  that looks for normal annotations `@app.route(...)` in its scope 

```scala
@flask def app: Flask = Flask()

@app.route("/")
def index() = "Index Page"

@app.route("/hello")
def hello() = "Hello, World"
```

it could then generate a `main` as follows
```scala
class app {
  <static> def main(args: Array[String]): Unit = {
    app
      .registerRoute("/", index)
      .registerRoute("/hello", hello)
}
```

Reply:
We might also have it as
```scala
@flask val app: Flask = Flask()
```
and modify it to
```scala
val app: Flask = 
  Flask()
    .registerRoute("/", index)
    .registerRoute("/hello", hello)
```

Reply:
Another approach that would follow closer the module initialization semantics of the Python version might be 
```scala
object app extends Flask {

  @route("/")
  def index() = "Index Page"

  @route("/hello")
  def hello() = "Hello, World"
}
```

where the `@root` macro annotation assumes it is inside a `Flask` instance and generates something like

```scala
object app extends Flask {

  val _ = (this: Flask).registerRoot("/" , index)
  def index() = "Index Page"

  val _ = (this: Flask).registerRoot("/hello", hello)
  def hello() = "Hello, World"
}
```

Reply:
I believe that macro annotations might not be the best tool to get an elegant solution to this particular problem. Based on these examples, __inline traits__ ([report](https://infoscience.epfl.ch/record/303591?ln=en), [prototype](https://github.com/scala/scala3/pull/17329)) are a much better fit. 

<details><summary>Cask with inline traits</summary>

```scala
object MinimalApplication extends MainAutoRoot {
  @get("/")
  def hello() = "Hello World!"

  @post("/do-thing")
  def doThing(request: Request) = request.text().reverse

  // generated code
  // def routes: Route = Seq(
  //   Route("hello", hello),
  //   Route("doThing", doThing, ...),
  // )
}
```

```scala
inline trait MainAutoRoot {
   def routes: Route = generateRoutes // macro `generateRoutes` finds the routes defined in the class that extends `Routes`
}

trait MainRoot {
  def roots: Roots
  def main(...) = // do something with routes
}
```

</details>  

> 1. Who defines the generateDefinition logic? Is it on the enclosing template? Or some property of bar or qux?

In an inline trait, it would be the trait itself.

> 2. If we need to generate multiple definitions, e.g. we have an object foo that contains both @cask.gets and MainArgs @main methods, how do we ensure the @cask.gets all get passed to Cask's generateDefinition logic, while all the MainArgs @mains get passed to MainArgs' generateDefinition logic?

With inline traits it becomes simpler because there is a single point of expansion for the definition. The definition expansion can collect all (non-macro) annotated definitions in scope.

Reply:
@nicolasstucki I think using an inline trait sounds reasonable. But if that's our stance, then that basically cuts out `Example 3: Main annotation` from the use cases of macro annotations: it doesn't make sense to say macro annotations provide a way to provide a worse version of MainArgs. We'll need to find other use cases to replace it, to make sure that macro annotations have enough motivating use cases to be worth it

Reply:
While it does not work for the particular API design choices of `MainArgs`, it allows for many other kinds of main annotation use cases. For example, it can be used to implement the generalized main annotation in https://github.com/scala/scala3/pull/19937.